src/controller.js

```diff
@@ -431,22 +448,43 @@ class Controller {
       throw new Error('Object must total fewer than 5 TB'); 
     }
 
-    let headers = this.__generateMetadataHeaders(metadata, {
+    let headers = {
       'content-sha256': sha256,
       'content-length': size,
-    });
+    };
+
+    // Figure out content-encoding and sha256 values
+    if (contentEncoding && contentEncoding !== 'identity') {
+      if (!transferSha256) {
+        throw new Error('When using Content-Encoding, a transferSha256 value must be given');
```

Comment:
Because the majority of implementations will not use `transferSha256` and I'd rather not force the implementation to specify the same hash twice in the same call.  The reason I have it structured like this is to make sure that when we need it, it's definitely there, but if we don't need it, we'll gladly accept a *valid* duplicated entry.  Basically, I'm thinking of a case where someone is iterating over something like 
```javascript
[
  {encoding: 'gzip', sha256: x, transferSha256: gzip(x)},
  {encoding: 'identity', sha256: y, transferSha256: identity(y)},
]
```
I think that setting the defaults and doing asserts later would be more confusing, personally.